pp_hot.c

```diff
@@ -5992,15 +5992,25 @@ Perl_leave_adjust_stacks(pTHX_ SV **from_sp, SV **to_sp, U8 gimme, int pass)
                  *    ++PL_tmps_ix, moving the previous occupant there
                  *    instead.
                  */
-                SV *newsv = newSV_type(SVt_NULL);
+
+                /* A newsv of type SVt_NULL will always be upgraded to
+                 * SvTYPE(sv), where that is a SVt_PVNV or below. It is
+                 * more efficient to create such types directly than
+                 * upgrade to them via sv_upgrade() within sv_setsv_flags. */
+                SV *newsv = (SvTYPE(sv) <= SVt_PVNV)
+                            ? newSV_type(SvTYPE(sv))
+                            : newSV_type(SVt_NULL);
 
                 PL_tmps_stack[++PL_tmps_ix] = *tmps_basep;
                 /* put it on the tmps stack early so it gets freed if we die */
                 *tmps_basep++ = newsv;
-
+#if NVSIZE <= IVSIZE
+                if (SvTYPE(sv) <= SVt_NV) {
+#else
                 if (SvTYPE(sv) <= SVt_IV) {
-                    /* arg must be one of undef, IV/UV, or RV: skip
-                     * sv_setsv_flags() and do the copy directly */
+#endif
```

Comment:
I don't know what our policy on preprocessor use is, but this could be written more compactly as
```c
if (SvTYPE(sv) <= (NVSIZE <= IVSIZE ? SVt_NV : SVt_IV)) {
```

Reply:
I'm certainly happy to write it like that. The current form is just mirroring what's already been in use in _Perl_sv_setsv_flags_.

Anyone else want to weigh in on preprocessor policy?

Reply:
I guess the only concern would be C compilers which can't do simple constant folding, which I'm assuming would be rare these days.

Reply:
The C standard requires constant folding.

>  A constant expression can be evaluated during translation rather than runtime, and accordingly may be used in any place that a constant may be.

In particular, you can use constant expressions as array sizes, `case` labels, bit-field sizes, and null pointer constants. If it doesn't constant fold, it's not a C compiler. :-)

Reply:
Ok, making those changes as suggested.

Reply:
Er, stuff has gone awry in tests, will have to look into it later.